File: model/economy.py
"""Simulates each team's credits round by round.

This exists because enemy credits are the single feature the model leans on
hardest, and they are never observable - not on screen, not in any API. They
can only be reconstructed by replaying the economy rules over the round
history, which is the same arithmetic experienced players do in their heads.

Drift is bounded rather than unbounded: credits reset to a known 800 at the
start of each half, so any accumulated error is wiped at halftime and can
never span a whole match.

MEASURED RESULT (21k rounds, 831 maps): this does not work well enough to
drive dollar-level advice, and the failure is structural rather than a matter
of tuning.

    mean absolute error   $5,924
    buy-bracket accuracy  35.1%  vs a 57.8% majority-class baseline

Three things defeat it, in order of severity:

1. Weapon carry-over. Survivors keep their guns for free, so a team holding
   $15k in credits can field a $25k loadout. Credits and loadout are simply
   different quantities, and the gap depends on who lived - which round
   outcomes do not tell you.
2. Spending is a choice. Two teams with identical credits and identical
   histories buy differently.
3. Kills pay $200 each, up to $1000 a round, invisibly.

The bias is systematic and one-directional: the simulator underestimates,
guessing semi-buy for 18,270 rounds that were actually full buys - exactly
what carry-over predicts.

What this means for the project: an automatic live tool cannot estimate enemy
economy from win/loss alone. It needs survivor counts per round, which are
visible on the round-end screen and therefore readable, plus kill counts. The
training pipeline is unaffected, because VLR reports real loadout values.

Kept rather than deleted: the measurement is the point, and it is what says
which inputs the live version actually requires.
"""

from dataclasses import dataclass, field

# Everything here is in TEAM credits - five players' worth - because that is
# the unit VLR reports, and comparing against it in per-player credits was the
# bug that made the first version of this look uncorrelated.
TEAM = 5

START_CREDITS = 800 * TEAM        # 4,000
OVERTIME_CREDITS = 5_000 * TEAM   # 25,000
MAX_CREDITS = 9_000 * TEAM        # 45,000

WIN_REWARD = 3_000 * TEAM
LOSS_BONUS = (1_900 * TEAM, 2_400 * TEAM, 2_900 * TEAM)
PLANT_REWARD = 300 * TEAM         # to attackers, paid even in a loss
KILL_REWARD = 200                 # per kill, not per team

REGULATION_ROUNDS = 24
HALFTIME_ROUND = 13


@dataclass
class TeamEconomy:
    credits: int = START_CREDITS
    loss_streak: int = 0

    def loss_bonus(self) -> int:
        return LOSS_BONUS[min(self.loss_streak, len(LOSS_BONUS) - 1)]


@dataclass
class EconomySimulator:
    """Replays a match's rounds and tracks both teams' credits.

    Credits are per player. Spending is not modelled: what a team actually
    bought is unobservable, so `spend()` is called explicitly when a loadout
    value is known, and left alone otherwise.
    """

    atk: TeamEconomy = field(default_factory=TeamEconomy)
    deff: TeamEconomy = field(default_factory=TeamEconomy)
    round_num: int = 1

    def _clamp(self, value: int) -> int:
        return max(0, min(MAX_CREDITS, value))

    def reset_half(self, overtime: bool = False) -> None:
        start = OVERTIME_CREDITS if overtime else START_CREDITS
        self.atk = TeamEconomy(credits=start)
        self.deff = TeamEconomy(credits=start)

    def spend(self, atk_spend: int = 0, def_spend: int = 0) -> None:
        self.atk.credits = self._clamp(self.atk.credits - atk_spend)
        self.deff.credits = self._clamp(self.deff.credits - def_spend)

    def resolve(
        self,
        winner_side: str,
        *,
        planted: bool = False,
        atk_kills: int = 0,
        def_kills: int = 0,
    ) -> None:
        """Apply one round's outcome and advance to the next round."""
        if winner_side not in ("atk", "def"):
            raise ValueError(f"winner_side must be 'atk' or 'def', got {winner_side!r}")

        winner, loser = (self.atk, self.deff) if winner_side == "atk" else (self.deff, self.atk)

        winner.credits = self._clamp(winner.credits + WIN_REWARD)
        loser.credits = self._clamp(loser.credits + loser.loss_bonus())

        winner.loss_streak = 0
        loser.loss_streak += 1

        # attackers are paid for a plant whether or not they go on to win
        if planted:
            self.atk.credits = self._clamp(self.atk.credits + PLANT_REWARD)

        self.atk.credits = self._clamp(self.atk.credits + atk_kills * KILL_REWARD)
        self.deff.credits = self._clamp(self.deff.credits + def_kills * KILL_REWARD)

        self.round_num += 1

        # sides swap at halftime and both economies reset to a known state
        if self.round_num == HALFTIME_ROUND:
            self.reset_half()
            self.atk, self.deff = self.deff, self.atk
        elif self.round_num == REGULATION_ROUNDS + 1:
            self.reset_half(overtime=True)

    def state(self) -> dict:
        return {
            "round": self.round_num,
            "atk_credits": self.atk.credits,
            "def_credits": self.deff.credits,
            "atk_loss_streak": self.atk.loss_streak,
            "def_loss_streak": self.deff.loss_streak,
        }


def buy_policy(credits: int) -> int:
    """How much a team is assumed to spend, given what it has.

    Real spending is unobservable, so this stands in for it: teams full-buy
    when they comfortably can, save hard when they clearly cannot, and force
    in between. Deliberately crude - `validate.py` measures what it costs.
    """
    if credits >= 20_000:          # comfortable full buy
        return int(credits * 0.85)
    if credits >= 12_000:          # force or half buy
        return int(credits * 0.75)
    return int(credits * 0.35)     # save, keeping most of it back


def simulate_match(rounds: list[dict]) -> list[dict]:
    """Replay stored rounds, returning the predicted economy before each one.

    `rounds` are dicts as produced by the scraper, ordered by round number.
    """
    sim = EconomySimulator()
    predicted = []

    for r in rounds:
        # what a team can field this round, before it spends anything
        pre_atk, pre_def = sim.atk.credits, sim.deff.credits
        # Spending cannot be read off loadout value, because survivors carry
        # weapons into the next round for free - loadout overstates spend by
        # whatever the team kept. So spend is modelled by policy instead.
        sim.spend(
            atk_spend=buy_policy(sim.atk.credits),
            def_spend=buy_policy(sim.deff.credits),
        )
        # bank is what VLR records: money left AFTER buying
        predicted.append({
            "round_num": r["round_num"],
            # pre-buy: comparable to loadout value / buy bracket
            "pred_atk_pre": pre_atk,
            "pred_def_pre": pre_def,
            # post-buy: comparable to VLR's bank figure
            "pred_atk_credits": sim.atk.credits,
            "pred_def_credits": sim.deff.credits,
        })
        sim.resolve(
            r["winner_side"],
            planted=r.get("win_condition") in ("spike", "defuse"),
        )

    return predicted

File: web/lib/economy.ts
/**
 * Tracks a match round by round and estimates both economies.
 *
 * Your own credits you can read off your screen. The enemy's you cannot -
 * they are not displayed anywhere, in game or in any API. So they are
 * reconstructed from round outcomes here.
 *
 * That reconstruction was measured against 21k real rounds and it is NOT
 * accurate to the dollar: $5,924 mean absolute error, because survivors keep
 * their weapons for free and loadout outruns credits by an amount round
 * outcomes never reveal. Two consequences shape this module:
 *
 *   1. The enemy figure is presented as an estimate you can correct, never
 *      as a fact.
 *   2. What is surfaced is the coarse bracket, not a precise number, because
 *      the bracket is what the win model consumes anyway.
 */

export const START_CREDITS = 800;
export const OVERTIME_CREDITS = 5_000;
export const MAX_CREDITS = 9_000;

export const WIN_REWARD = 3_000;
export const LOSS_BONUS = [1_900, 2_400, 2_900] as const;
export const PLANT_REWARD = 300;

export const HALFTIME_ROUND = 13;
export const REGULATION_ROUNDS = 24;

export type RoundOutcome = "won" | "lost" | "lost-planted" | "won-planted";

export interface RoundRecord {
  round: number;
  outcome: RoundOutcome;
  youAttacking: boolean;
}

export interface MatchState {
  map: string;
  /** Which side you started the match on. */
  startedAttacking: boolean;
  history: RoundRecord[];
  /** Your own credits, read off your screen; null means "use the estimate". */
  yourCreditsOverride: number | null;
  /** Manual correction when you can see the estimate is wrong. */
  enemyCreditsOverride: number | null;
}

export function newMatch(map: string, startedAttacking: boolean): MatchState {
  return {
    map,
    startedAttacking,
    history: [],
    yourCreditsOverride: null,
    enemyCreditsOverride: null,
  };
}

export function currentRound(state: MatchState): number {
  return state.history.length + 1;
}

export const ROUNDS_TO_WIN = 13;

/**
 * Sides swap at halftime, and then every single round in overtime, so that
 * each team plays one attack and one defence per overtime pair.
 */
export function youAreAttacking(state: MatchState): boolean {
  const round = currentRound(state);
  if (round > REGULATION_ROUNDS) {
    const overtimeIndex = round - REGULATION_ROUNDS - 1;
    return overtimeIndex % 2 === 0 ? state.startedAttacking : !state.startedAttacking;
  }
  const secondHalf = round >= HALFTIME_ROUND;
  return secondHalf ? !state.startedAttacking : state.startedAttacking;
}

export interface MatchResult {
  winner: "you" | "them";
  you: number;
  them: number;
  wentToOvertime: boolean;
}

/**
 * Whether the match is over, and who took it.
 *
 * Regulation is first to 13. At 12-12 it goes to overtime, which is played in
 * pairs of rounds - a team wins by taking both, which is the same as leading
 * by two once a pair has finished. A lead of two mid-pair does not end it,
 * because the other team has not had its round on the opposite side yet.
 */
export function matchResult(state: MatchState): MatchResult | null {
  const { you, them } = score(state);
  const played = state.history.length;

  if (played <= REGULATION_ROUNDS) {
    // Reaching 13 inside regulation always wins it: 12-12 sends the match to
    // overtime at 24 rounds, so a 13-12 regulation score cannot occur.
    if (you >= ROUNDS_TO_WIN) return { winner: "you", you, them, wentToOvertime: false };
    if (them >= ROUNDS_TO_WIN) return { winner: "them", you, them, wentToOvertime: false };
    return null;
  }

  // overtime: only judge at the end of a completed pair
  const overtimeRounds = played - REGULATION_ROUNDS;
  if (overtimeRounds % 2 !== 0) return null;

  const lead = you - them;
  if (lead >= 2) return { winner: "you", you, them, wentToOvertime: true };
  if (lead <= -2) return { winner: "them", you, them, wentToOvertime: true };
  return null;
}

export function score(state: MatchState): { you: number; them: number } {
  let you = 0;
  let them = 0;
  for (const r of state.history) {
    if (r.outcome === "won" || r.outcome === "won-planted") you++;
    else them++;
  }
  return { you, them };
}

function clamp(v: number): number {
  return Math.max(0, Math.min(MAX_CREDITS, v));
}

interface Economy {
  you: number;
  them: number;
  yourLossStreak: number;
  theirLossStreak: number;
}

/**
 * How much a team is assumed to spend. Real spending is a choice nobody can
 * observe, so this stands in for it - deliberately simple, because measurement
 * showed no policy recovers the true number.
 */
function assumedSpend(credits: number): number {
  if (credits >= 3_900) return 3_900; // rifle plus shields
  if (credits >= 2_400) return 2_200; // half buy
  if (credits >= 1_600) return 1_400; // pistol plus shields
  return 400;
}

export function estimateEconomy(state: MatchState): { you: number; them: number } {
  let eco: Economy = {
    you: START_CREDITS,
    them: START_CREDITS,
    yourLossStreak: 0,
    theirLossStreak: 0,
  };

  state.history.forEach((record, index) => {
    const roundNumber = index + 1;

    eco.you = clamp(eco.you - assumedSpend(eco.you));
    eco.them = clamp(eco.them - assumedSpend(eco.them));

    const youWon = record.outcome === "won" || record.outcome === "won-planted";
    const planted = record.outcome === "lost-planted" || record.outcome === "won-planted";

    if (youWon) {
      eco.you = clamp(eco.you + WIN_REWARD);
      eco.them = clamp(eco.them + LOSS_BONUS[Math.min(eco.theirLossStreak, 2)]);
      eco.yourLossStreak = 0;
      eco.theirLossStreak += 1;
    } else {
      eco.them = clamp(eco.them + WIN_REWARD);
      eco.you = clamp(eco.you + LOSS_BONUS[Math.min(eco.yourLossStreak, 2)]);
      eco.theirLossStreak = 0;
      eco.yourLossStreak += 1;
    }

    // the plant bonus is paid to whoever was attacking that round
    if (planted) {
      if (record.youAttacking) eco.you = clamp(eco.you + PLANT_REWARD);
      else eco.them = clamp(eco.them + PLANT_REWARD);
    }

    // both economies reset to a known state at halftime, which is what stops
    // estimation error from accumulating across a whole match
    const next = roundNumber + 1;
    if (next === HALFTIME_ROUND) {
      eco = { you: START_CREDITS, them: START_CREDITS, yourLossStreak: 0, theirLossStreak: 0 };
    } else if (next === REGULATION_ROUNDS + 1) {
      eco = { you: OVERTIME_CREDITS, them: OVERTIME_CREDITS, yourLossStreak: 0, theirLossStreak: 0 };
    }
  });

  return {
    you: state.yourCreditsOverride ?? eco.you,
    them: state.enemyCreditsOverride ?? eco.them,
  };
}

/** Rounds since the last known-exact economy: round 1, halftime, overtime. */
export function roundsSinceReset(state: MatchState): number {
  const round = currentRound(state);
  if (round > REGULATION_ROUNDS) return (round - REGULATION_ROUNDS - 1) % 6;
  if (round >= HALFTIME_ROUND) return round - HALFTIME_ROUND;
  return round - 1;
}

/**
 * Confidence in the enemy estimate, which decays with distance from the last
 * reset. Stated honestly rather than hidden: the estimate is known to drift.
 */
export function estimateConfidence(state: MatchState): "exact" | "good" | "rough" {
  if (state.enemyCreditsOverride !== null) return "exact";
  const since = roundsSinceReset(state);
  if (since === 0) return "exact";
  return since <= 3 ? "good" : "rough";
}
